Refactor build_cpe.py

jan-cerny opened the pull request:
This PR adds some style improvements proposed in #9799:
* Move function `is_parametrized` to `Symbol` class
* Extract function `recurse_or_substitute_or_do_nothing`

For more details, please read commit messages of each commit.

matejak commented on ssg/boolean_expression.py:
```diff
@@ -112,6 +112,10 @@ def version_definitions(self):
         return self.requirement.project_name
 
+    @staticmethod
+    def _is_parametrized(ref):
```
So now the method shouldn't be private, as you intend it to be used by other callers, and `ref` doesn't give a hint about what is expected as parameter. In fact, the naming and usage `Symbol.is_parametrized` is misleading, because it's not the symbol and its parametrization that you are evaluating - instead, you are looking at a string that represents a platform unit. So please give a thought about a name of the method and of the argument as well.

matejak commented on ssg/utils.py:
```diff
@@ -342,13 +351,9 @@ def apply_formatting_on_dict_values(source_dict, string_dict, ignored_keys=[]):
-                new_dict[k] = v.format(**string_dict)
-            else:
-                new_dict[k] = v
+        if ignored_keys is None or k not in ignored_keys:
```
It's better to make `ignored_keys` an empty list/set if it is not given, and to do so at the top of the function. That way, the `if` statement may be made human-readable.

matejak reviewed (changes requested):
Moves in the right direction [towards the project guidelines](https://github.com/ComplianceAsCode/content/blob/master/docs/manual/developer/04_style_guide.md#general-coding-style), but it is still possible to improve with little effort. I have pointed out a couple of simple changes.

evgenyz commented on ssg/utils.py:
```diff
@@ -333,7 +333,16 @@ def enum(*args):
 
 
-def apply_formatting_on_dict_values(source_dict, string_dict, ignored_keys=[]):
+def recurse_or_substitute_or_do_nothing(v, string_dict, ignored_keys):
```
```suggestion
def recurse_or_substitute_or_do_nothing(v, string_dict, ignored_keys=frozenset()):
```
This way you won't have to check for None. It'll retain the iterable type and won't be affected by the "mutable default argument" problem.

jan-cerny commented:
I have rebased this PR on the top of the latest upstream master branch. I have used `frozenset` as a default argument, I have renamed method `_is_parametrized` to `reference_is_parametrized` and then set `ignored_keys` to an empty list explicitly.

jan-cerny commented on ssg/utils.py:
```diff
@@ -333,22 +333,30 @@ def enum(*args):
+        return v.format(**string_dict)
+    else:
+        return v
+
```
@evgenyz Where the frozenset should be after the refactoring to make it work as expected? Should it be in `recurse_or_substitute_or_do_nothing` or in the `apply_formatting_on_dict_values`?

matejak replied:
Preferably in both places, so the default argument to `ignored_keys` is an empty iterable that is not mutable. That way, you even won't have to set `ignored_keys` to that empty list.
However, the decomposition into those two functions in this form was not such great idea, because the `recurse_or_substitute_or_do_nothing` accepts `ignored_keys` that it doesn't care about - it just passes those on, and it also hardcodes "recursion" to calling `apply_formatting_on_dict_values`, which creates a circular dependency. So I would say that having seen this, I would either revert to the original code, or if you became sold to the idea of decomposing handling of container and handling of items, I would pass `recurse_or_substitute_or_do_nothing` a callback, e.g. `lambda value, substitutions: recurse_or_substitute_or_do_nothing(value, substitutions, ignored_keys)` that would be used in the item-processing function if the item passed is a dictionary.
So sorry, suggesting the decomposition like this was a mistake.

evgenyz commented on ssg/utils.py:
```diff
@@ -333,22 +333,30 @@ def enum(*args):
     """
     new_dict = {}
+    if ignored_keys is None:
+        ignored_keys = []
```
When `ignored_keys=frozenset()` these 2 lines won't be needed.

matejak commented on ssg/boolean_expression.py:
```diff
@@ -93,6 +93,10 @@ def arg(self):
         return self.requirement.project_name
 
+    @staticmethod
+    def reference_is_parametrized(reference):
```
The problem with naming still remains - when considering that `Symbol` class, what is supposed to be this `reference` argument that the method accepts?

jan-cerny commented:
I have renamed a variable ref, I have also renamed the method `reference_is_paramaterized` to `cpe_id_is_parametrized` and changed `ignored_keys` to a `frozenset`.

vojtapolasek commented on ssg/build_cpe.py:
```diff
@@ -84,19 +84,17 @@ def add_cpe_item(self, cpe_item):
         return ref.startswith("cpe:")
 
-    def _is_parametrized(self, ref):
-        return re.search(r'^\w+\[\w+\]$', ref)
```
By applying this change, the re module is no longer needed and does not need to be imported.

jan-cerny commented on ssg/build_cpe.py:
```diff
@@ -84,19 +83,17 @@ def add_cpe_item(self, cpe_item):
         try:
-            if self._is_name(ref):
-                return self.cpes_by_name[ref]
+            if self._is_name(cpe_id_or_name):
```
@matejak Should the `_is_name` be moved to Symbol class as well?

matejak replied:
Good catch, the class `ProductCPEs` definitely isn't the one that has the knowledge about how data structures format stuff themselves.
The naming `is_name` is also confusing - for example, to me "Jan" is a name, whereas `human:Jan` is not what I would identify as name, it's perhaps some sort of an ID.

jan-cerny commented:
I have removed the unused re import and I have move the is_name method to the Symbol class.

matejak commented:
The renaming is a huge success, thanks!
